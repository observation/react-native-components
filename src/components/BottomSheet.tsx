import React from 'react'
import { StyleSheet, View, Text, ViewStyle, StyleProp } from 'react-native'

import { useTheme } from '@react-navigation/native'

import LargeButton, { LargeButtonProps } from './LargeButton'
import textStyle from '../styles/text'
import theme from '../styles/theme'

type Props = {
  title?: string
  text?: string
  buttons?: LargeButtonProps[]
  style?: StyleProp<ViewStyle>
  testID?: string
  children?: React.ReactNode
}

const BottomSheet = ({ title, text, buttons = [], style, testID, children }: Props) => {
  const { colors } = useTheme()
  const buttonsMarginTop = title || text ? theme.margin.common : 0
  return (
    <View style={[styles.container, style]} testID={testID}>
      <View style={[styles.bottomSheetContainer, { backgroundColor: colors.card }]}>
        <View style={styles.bottomSheet}>
          <View style={{ flexDirection: 'row', alignItems: 'flex-start' }}>
            {title && (
              <View style={{ flex: 1 }}>
                <Text style={textStyle.lead}>{title}</Text>
              </View>
            )}
          </View>
          {text && (
            <View style={{ marginTop: theme.margin.half }}>
              <Text style={textStyle.body}>{text}</Text>
            </View>
          )}
          {children}
          {buttons && buttons.length > 0 && (
            <View
              style={{
                marginTop: buttonsMarginTop,
                marginHorizontal: -theme.margin.half,
                flexDirection: 'row',
                alignItems: 'flex-start',
              }}
            >
              {buttons.map((button) => (
                <LargeButton
                  iconName={button.iconName}
                  disabled={button.disabled}
                  secondary={button.secondary}
                  title={button.title}
                  key={button.title}
                  style={styles.buttonStyle}
                  onPress={button.onPress}
                />
              ))}
            </View>
          )}
        </View>
      </View>
    </View>
  )
}

export default BottomSheet

const styles = StyleSheet.create({
  container: {
    ...theme.shadow.ios,
    borderTopWidth: 1 / 3,
    borderTopColor: theme.color.greySemi,
  },
  bottomSheetContainer: {
    ...theme.shadow.android,
  },
  bottomSheet: {
    flexDirection: 'column',
    margin: theme.margin.common,
  },
  buttonStyle: {
    flex: 1,
    margin: 0,
    marginHorizontal: theme.margin.half,
  },
})
