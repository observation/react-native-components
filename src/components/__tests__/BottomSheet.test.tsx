import React from 'react'
import { Text } from 'react-native'

import { fireEvent, render } from '@testing-library/react-native'

import BottomSheet from '../BottomSheet'

describe('BottomSheet', () => {
  describe('Rendering', () => {
    test('Without icon', () => {
      // GIVEN
      const { toJSON } = render(<BottomSheet title="Title" text="Text" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With button', () => {
      // GIVEN
      const { toJSON } = render(
        <BottomSheet title="Title" text="Text" buttons={[{ onPress: () => {}, title: 'Action!' }]} />,
      )
      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Only buttons', () => {
      // GIVEN
      const { toJSON } = render(<BottomSheet buttons={[{ onPress: () => {}, title: 'Action!' }]} />)
      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With children', () => {
      // GIVEN
      const { queryByText, toJSON } = render(
        <BottomSheet>
          <Text>Hello world!</Text>
        </BottomSheet>,
      )

      // THEN
      expect(queryByText('Hello world!')).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })
  })

  describe('Interaction', () => {
    test('Click on a button', () => {
      // GIVEN
      const onPress = jest.fn()
      const { getByText } = render(<BottomSheet title="Title" text="Text" buttons={[{ onPress, title: 'Action!' }]} />)

      // WHEN
      fireEvent.press(getByText('Action!'))

      // THEN
      expect(onPress).toBeCalled()
    })
  })
})
