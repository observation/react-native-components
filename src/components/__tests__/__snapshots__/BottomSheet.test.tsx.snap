// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet Rendering Only buttons 1`] = `
<View
  style={
    [
      {
        "borderTopColor": "#E6E6E6",
        "borderTopWidth": 0.3333333333333333,
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "elevation": 6,
          "shadowColor": "#939393",
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "marginHorizontal": -8,
            "marginTop": 0,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#0066B1",
              "borderRadius": 4,
              "flex": 1,
              "height": 32,
              "justifyContent": "center",
              "margin": 0,
              "marginHorizontal": 8,
              "opacity": 1,
              "overflow": "hidden",
            }
          }
          testID="touchable-opacity"
        >
          <View
            style={
              {
                "flexDirection": "row",
                "marginHorizontal": 16,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#212121",
                    "fontFamily": "Ubuntu",
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "bold",
                    "lineHeight": 20,
                    "textAlignVertical": "center",
                  },
                  {
                    "color": "#FFFFFF",
                  },
                ]
              }
            >
              Action!
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`BottomSheet Rendering With button 1`] = `
<View
  style={
    [
      {
        "borderTopColor": "#E6E6E6",
        "borderTopWidth": 0.3333333333333333,
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "elevation": 6,
          "shadowColor": "#939393",
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "marginHorizontal": -8,
            "marginTop": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#0066B1",
              "borderRadius": 4,
              "flex": 1,
              "height": 32,
              "justifyContent": "center",
              "margin": 0,
              "marginHorizontal": 8,
              "opacity": 1,
              "overflow": "hidden",
            }
          }
          testID="touchable-opacity"
        >
          <View
            style={
              {
                "flexDirection": "row",
                "marginHorizontal": 16,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#212121",
                    "fontFamily": "Ubuntu",
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "bold",
                    "lineHeight": 20,
                    "textAlignVertical": "center",
                  },
                  {
                    "color": "#FFFFFF",
                  },
                ]
              }
            >
              Action!
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`BottomSheet Rendering With children 1`] = `
<View
  style={
    [
      {
        "borderTopColor": "#E6E6E6",
        "borderTopWidth": 0.3333333333333333,
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "elevation": 6,
          "shadowColor": "#939393",
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      />
      <Text>
        Hello world!
      </Text>
    </View>
  </View>
</View>
`;

exports[`BottomSheet Rendering Without icon 1`] = `
<View
  style={
    [
      {
        "borderTopColor": "#E6E6E6",
        "borderTopWidth": 0.3333333333333333,
        "shadowColor": "#E6E6E6",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "elevation": 6,
          "shadowColor": "#939393",
        },
        {
          "backgroundColor": "rgb(255, 255, 255)",
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "margin": 16,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#212121",
                "fontFamily": "Ubuntu",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 20,
              }
            }
          >
            Title
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "marginTop": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontFamily": "Ubuntu",
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "lineHeight": 20,
            }
          }
        >
          Text
        </Text>
      </View>
    </View>
  </View>
</View>
`;
